doc/coding-guidelines.md

```diff
@@ -0,0 +1,447 @@
+
+
+## CSS
+<a name="css-use-lower-case-with-dashes"></a>
+* [1.](#css-use-lower-case-with-dashes) Use the `lower-case-with-dashes` format.
+<a name="css-prefix-global-classes"></a>
+* [2.](#css-prefix-global-classes) Prefix classes with `theia` when used as global classes.
+<a name="no-styles-in-code"></a>
+* [3.](#no-styles-in-code) Do not define styles in code. Introduce proper CSS classes.
+> Why? It is not possible to play with such styles in the dev tools without recompiling the code. CSS classes can be edited in the dev tools.
+
+## Theming
+<a name="theming-no-css-color-variables"></a>
+* [1.](#theming-no-css-color-variables) Do not introduce css color variables. Implement `ColorContribution` and use `ColorRegistry.register` to register new colors.
+<a name="theming-no-css-color-values"></a>
+* [2.](#theming-no-css-color-values) Do not introduce hardcode color values in css. Instead reference [VS Code colors](https://code.visualstudio.com/api/references/theme-color) in css by prefixing them with `--theia` and replacing all dots with dashes. For example `widget.shadow` color can be referenced in css with `var(--theia-widget-shadow)`.
+<a name="theming-derive-colors-from-vscode"></a>
+* [3.](#theming-derive-colors-from-vscode) Always derive new colors from existing [VS Code colors](https://code.visualstudio.com/api/references/theme-color). One can derive from an existing color, just by plainly referencing it, e.g. `dark: 'widget.shadow'`, or applying transformations, e.g. `dark: Color.lighten('widget.shadow', 0.4)`. 
+> Why? Otherwise, there is no guarantee that new colors don't look alien for a random VS Code color theme.
+<a name="theming-theia-colors"></a>
+* [4.](#theming-theia-colors) Apply different color values only in concrete Theia themes, see [Light (Theia)](https://github.com/eclipse-theia/theia/blob/master/packages/monaco/data/monaco-themes/vscode/light_theia.json), [Dark (Theia)](https://github.com/eclipse-theia/theia/blob/master/packages/monaco/data/monaco-themes/vscode/dark_theia.json) and [High Contrast (Theia)](https://github.com/eclipse-theia/theia/blob/master/packages/monaco/data/monaco-themes/vscode/hc_theia.json) themes.
+<a name="theming-variable-naming"></a>
+* [5.](#theming-variable-naming) Names of variable follow the `object.property` pattern.
+```ts
+// bad
+'button.secondary.foreground'
+'button.secondary.disabled.foreground'
+
+// good
+'secondaryButton.foregrond'
+'secondaryButton.disabledForegrond'
+```
+
+## React
+<a name="no-bind-fn-in-event-handlers"></a>
+* [1.](#no-bind-fn-in-event-handlers) Do not bind functions in event handlers.
+  - Extract a react component if you want to pass state to an event handler function.
+
+> Why? Because it creates a new instance of event handler function on each render and breaks React element caching leading to rerendering and bad performance.
```

Comment:
```suggestion
> Why? Because doing so creates a new instance of the event handler function on each render and breaks React element caching leading to rerendering and bad performance.
```